design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,315 @@
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type RejectedMigratedVolume struct {
+	SourcePvc string `json:"sourcePvc,omitempty" valid:"required"`
+	Reason    string `json:"reason,omitempty" valid:"required"`
+}
+
+type RejectedVolumeMigration struct {
+	RejectedVolumes []RejectedMigratedVolume `json:"rejectedVolumes,omitempty" valid:"required"`
+	AssociatedVolumes []string `json:"associatedVolumes,omitempty"`
+	Reason            string   `json:"reason,omitempty" valid:"required"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRejected   = "Rejected"
+	VolumeMigrationPhaseRunning    = "Running"
+	VolumeMigrationPhaseScheduling = "Scheduling"
+	VolumeMigrationPhaseSucceeded  = "Succeeded"
+	VolumeMigrationPhaseUnknown    = "Unknown"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type VolumeMigrationState struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+	VirtualMachineMigrationName string `json:"virtualMachineMigrationName,omitempty"`
+	Succeeded bool `json:"succeeded,omitempty"`
+	Failed bool `json:"failed,omitempty"`
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+	EndTimestamp *metav1.Time `json:"endTimestamp,omitempty"`
+}
+
+type VolumeMigrationStatus struct {
+	VirtualMachineInstanceName string                  `json:"virtualMachineInstanceName,omitempty"`
+	VolumeMigrationState       VolumeMigrationState    `json:"volumeMigrationState,omitempty"`
+	PendingVolumes             []MigratedVolume        `json:"pendingVolumes,omitempty"`
+	RejectedVolumeMigration    RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`
+	Phase                      VolumeMigrationPhase    `json:"phase,omitempty"`
+	// PhaseTransitionTimestamp is the timestamp of when the last phase change occurred
+	PhaseTransitionTimestamps []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
+}
+```
+
+
+Potentially, users could specify multiple migrated volumes and those could belong to the same or different VMI or none of the VMIs. The volume migration controller will classify and group the volumes defined on the same VMI.
+
+Currently, this API supports to migrate PVCs only if they are associated to a running VM. Therefore, the controller will reject all the volume migration objects that have migrated volumes belonging to different VMIs or aren't associated to any. If all the migrated volumes aren't associated to a VMI, the object is in `pending` state as the VMI could be created in a later time.
+
+The `VolumeMigration` expects a list of source and destination PVCs and the destination PVCs needs to be created either manually by the user or other tools.
```

Comment:
One thing that bothers me here is: how do I know which node the destination PVCs should be bound to? If I understand correctly, in order to prepare for the volume migration I need to somehow find a node that have sufficient amount of resources (CPU, memory, devices, storage, etc.) to accommodate my VM. Only then I should create the PVCs so that they are assigned to that node. Does that mean that users will need to 'predict' where the scheduler will migrate the VM to? Or maybe do not use PVCs with immediate binding (only WFFC)? Or how is that supposed to work?

Reply:
@vasiliy-ul I think this a general problem with immediate binding PVCs not specific to this design. It is the case also if you create a VM with a Immediate binding PVC. Potentially, the virt-launcher pod could remain in pending state if the node where the PVC is bound doesn't match the scheduling constraints.

Reply:
So remember, PVCs are not bound to nodes (unless it is a local volume with a WFFC storage) immediate binding PVCs in general are just bound to the matching PV, and then when the workload is scheduled to run on a node, the volume is attached to that node before the workload starts running. But that has nothing to do with the status of the volume.

Reply:
:facepalm: Right, I was probably thinking too much about it with local volumes in mind.